test/test_images/probe_helper/utils/liveness.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	nethttp "net/http"
+	"time"
+
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+)
+
+// LivenessChecker carries timestamps of the latest handled events from the
+// forwarder and receiver, as well as the longest tolerated staleness time.
+// The LivenessChecker fails checks if the probe helper is unable to handle
+// either sort of event.
+type LivenessChecker struct {
+	// LastForwardEventTime is the timestamp of the last event processed by the forward client.
+	LastForwardEventTime SyncTime
+
+	// LastReceiverEventTime is the timestamp of the last event processed by the receiver client.
+	LastReceiverEventTime SyncTime
+
+	// LivenessStaleDuration represents the tolerated staleness of events processed by the forward and receiver clients in the liveness probe.
+	LivenessStaleDuration time.Duration
+
+	// SchedulerStaleDuration represents the tolerated staleness of ticks observed by the CloudSchedulerSource probe.
+	SchedulerStaleDuration time.Duration
```

Comment:
I'd like to avoid this dependency if possible. I suggest that the LivenessChecker maintain a list of "ActionFuncs" to run upon each liveness check. Again, this is about isolation: LivenessChecker shouldn't have knowledge about scheduler source. But it's OK for it to run a generic "ActionFunc".

Reply:
Done.